[51] Binary Classifier training on the GPT labelled data

j-gillam opened the pull request:
---

# Description

This addresses Issue [51](https://github.com/nestauk/discovery_child_development/issues/51). 
Included in this PR:

- Moved folders for 01-06 from the openalex/binary_classifier to openalex/openalex_binary_classifier
- Added pipeline for the gpt labelled training set pipeline to the models folder in pipeline. (The code is the same for this pipeline, with very minor alterations to use this dataset instead)
- Adds functionality to download trained model from S3 to test new datasets with using tar.gz files as it wouldn't save them a pickle files for huggingface models. 
- 05_evaluating_openalex_data.py uses the new model to predict relevance labels for a random sample of relevant/not relevant openalex works. Plus saves the dataset with predictions. 

To review this PR:

- Have a go at running the notebooks, specifically the 05 one! It will run a lot faster in skypilot, so if you did want to do that, change to the branch in the task.yaml! 

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

j-gillam committed:
Pipeline for labelled data

j-gillam committed:
Pipeline for labelled data

j-gillam committed:
Pipeline for labelled data

j-gillam committed:
Updating notebooks

j-gillam committed:
Updating notebooks

j-gillam committed:
Evaluating labelled data model

j-gillam committed:
Evaluating labelled data model

j-gillam committed:
Evaluating labelled data model

j-gillam committed:
Evaluating labelled data model

j-gillam committed:
Merge branch '41-classifier-pipeline' into 51-binary-classifier-gpt-labelled

j-gillam committed:
Evaluating labelled data model

j-gillam committed:
Adding breakdown of results by source

j-gillam committed:
Adding breakdown of results by source

j-gillam committed:
Adding breakdown of results by source

j-gillam committed:
Bugfix

j-gillam committed:
Bugfix

j-gillam committed:
Bugfix

j-gillam committed:
Adding changes from PR 49

j-gillam committed:
Adding changes from PR 49

j-gillam committed:
Removing outputs from notebooks

j-gillam committed:
Adding fixes from PR 49

j-gillam committed:
Adding fixes from PR 49

j-gillam committed:
Adding crunchbase results for GPT labelled model

j-gillam committed:
Updating save path

j-gillam committed:
Updating save path

j-gillam committed:
Updating save path

RFOxbury commented on discovery_child_development/config/config.yaml:
```diff
@@ -62,3 +62,4 @@ patents_data_path: "data/patents/"
 patents_data_version: "GooglePatents_20231106_175151"
 
 embedding_sample_size: 1000
+labelled_sample_size: 50
```
How come you only use 50? Also, as with the previous PR, I see that this value gets read in as `NUM_SAMPLES` but I can't figure out where that is used! What is it doing?

j-gillam replied:
Good question. It is used in in models/binary_classifier/03_... To create a non-production example, for testing purposes. And as we only have around 100 samples in the validation set, I just arbitrarily chose 50 for the training/validation to test the pipeline was working.

RFOxbury commented on discovery_child_development/notebooks/binary_classifier/04b_train_distilbert_classifier_gpt_labelled.ipynb:
With the last part of the notebook, "7. Trialling the model on the openalex concepts", this is how I've understood it:

- Karlis provided some patents that had been labelled by GPT
- The earlier part of this notebook fine-tunes a distilbert model on a sample of these GPT-labelled patents, with a validation set of 50
- In this final section of the notebook, you're loading in some relevant and not relevant OpenAlex works to see how the fine-tuned distilbert model performs on them

Is that correct?

j-gillam replied:
Very close!

'The earlier part of this notebook fine-tunes a distilbert model on a sample of these GPT-labelled patents, with a validation set of' 10% of the gpt labels. The 50 was only used for the non-production test example. Thats my bad I forgot to remove it from all the non-embedding files.

j-gillam replied:
But actually a lot more of the evaludation (plus some in this file) goes on in 05b and testing the saving and redownloading of the models.

RFOxbury commented on discovery_child_development/pipeline/models/binary_classifier/04b_train_distilbert_classifier.py:
```diff
@@ -0,0 +1,203 @@
+            project="ISS supervised ML",
+            job_type="Binary classifier - huggingface",
+            save_code=True,
+            tags=["huggingface", "binary_classifier", "sentence_embeddings", "all"],
```
This is a really minor thing but seeing as the job type is "binary classifier", I don't think you need this in the tags as well?

Also what does the "sentence_embeddings" tag mean? The simple classifiers are trained on sentence embeddings from the all-miniLM model, whereas this code uses embeddings created with distilbert, right? Or is that actually wrong?

j-gillam replied:
No clue really! Will remove haha, I might just specify where it is miniLM vs distilbert

RFOxbury commented on discovery_child_development/pipeline/models/binary_classifier/05_evaluating_openalex_data.py:
```diff
@@ -0,0 +1,169 @@
+    labelled_data_ids = labelled_data.id.unique()
+
+    # Get abstracts
+    abstracts = get_abstracts().query("id not in @labelled_data_ids")
```
I did not know you could do `.query()` and `@` the object! That's cool

RFOxbury reviewed (approved):
Hey Jess, this looks good and seems to work! I got confused by some of the details, eg. if in notebook `04b_train_distilbert_classifier_gpt_labelled.ipynb` you are fine-tuning a pre-trained model, or if that model was already fine-tuned (on the last PR) and you're training it some more. I think it's the former ie we're not using a model from the last PR, but wanted to check. In general, in the notebooks, would you be able to add a line at the top of some of the more complicated sections saying what is going on? So that tired brains like me don't get confused!

Note: I have not tried running the pipeline scripts, just running the notebook equivalents

j-gillam committed:
Fixes for PR 41

j-gillam committed:
Edits based on PR 51

j-gillam committed:
Merge branch '41-classifier-pipeline' into 51-binary-classifier-gpt-labelled

j-gillam commented:
> Hey Jess, this looks good and seems to work! I got confused by some of the details, eg. if in notebook `04b_train_distilbert_classifier_gpt_labelled.ipynb` you are fine-tuning a pre-trained model, or if that model was already fine-tuned (on the last PR) and you're training it some more. I think it's the former ie we're not using a model from the last PR, but wanted to check. In general, in the notebooks, would you be able to add a line at the top of some of the more complicated sections saying what is going on? So that tired brains like me don't get confused!
> 
> Note: I have not tried running the pipeline scripts, just running the notebook equivalents

Hey @RFOxbury I have added some extra text and made the edits from both the PR's. I have added some info to the top of each notebook but can add extra text in the new year! I am going to merge now in case you need it before I get back!
